### ISSUE TITLE
modularize MoleculeModel

### ISSUE THREAD

**Issue by davidegraff** [chemprop/chemprop#43]
Hello!

Just wanted to start off by saying fantastic work on this project- it's truly an exciting advancement in molecular property prediction!

I'm writing to ask about the possibility of modularizing the Chemprop model. I'm trying to incorporate the model into some software of my own, but am finding it pretty unnatural to do so. My main issue arises from the need to extensively wrap code you've already written.
ex: The central model in Chemprop, the MoleculeModel class, takes "one" argument, args, of type TrainArgs. Essentially, this is just a Namespace that holds all of the keyword arguments to properly initialize a MoleculeModel, but it ends up obfuscating the arguments required to initialize a MoleculeModel. Additionally, one can't just initialize a TrainArgs object and pass that, because that requires specifying some unnecessary arguments (e.g. 'data_path' and 'target_columns') that are irrelevant to just creating the MoleculeModel itself. I got around this by manually inspecting the MoleculeModel class and searching for what attributes of a TrainArgs object that it accesses during initialization, then cross-referencing these with their default values from TrainArgs to create a wrapped initializer for a MoleculeModel, but I think your code would benefit from directly incorporating such an initializer so that future users like myself won't have to go to such lengths. (See below for a sample implementation of this. Also, please correct me if I've gotten anything wrong!)

```python
class MPNN:
    """A wrapper for the base chemprop MoleculeModel"""
    def __init__(self, batch_size: int = 50,
                 atom_messages: bool = False, hidden_size: int = 300,
                 bias: bool = False, depth: int = 3, dropout: float = 0.0,
                 undirected: bool = False, features_only: bool = False,
                 use_input_features: bool = False, device: str = 'cpu',
                 features_size: Optional[int] = None, activation: str = 'ReLU',
                 ffn_hidden_size: Optional[int] = None,
                 ffn_num_layers: int = 2, num_tasks: int = 1,
                 metric: str = 'rmse', epochs: int = 30,
                 warmup_epochs: float = 2.0, total_epochs=0,
                 init_lr: float = 1e-4, max_lr: float = 1e-3,
                 final_lr: float = 1e-4):
        ffn_hidden_size = ffn_hidden_size or hidden_size
        self.model = MoleculeModel(Namespace(
            dataset_type='regression', num_tasks=num_tasks,
            atom_messages=atom_messages, hidden_size=hidden_size,
            bias=bias, depth=depth, dropout=dropout,
            undirected=undirected, features_only=features_only,
            use_input_features=use_input_features, device=device,
            features_size=features_size, activation=activation,
            ffn_hidden_size=ffn_hidden_size, ffn_num_layers=ffn_num_layers), featurizer=False
        )
```

edit: I now disagree with the point from my original comment reproduced below
> Related to this (this is solely an opinion)- what are your thoughts on creating wrapped train() and predict() methods for Chemrop to more easily inline these methods into other software? My motivation for asking is that other types of predictive models like those in sklearn and tensorflow offer pretty straightforward syntax for fitting a model based on two parallel sequences of inputs and outputs. Trying to do the same for the Chemprop model is also proving to be not as trivial as one might hope. I think it might be beneficial for others looking to use your model in their own software if you provide these simplified methods as part of your library.

**Comment by swansonk14** [comment]
Hi @davidegraff, thank you so much for the suggestions and for going to all the work to create the wrapper!

I agree with your design in many ways and in older versions of Chemprop we had implemented it that way, but the problem we ran up against was that making the arguments to the various models and functions explicit made it much more difficult to rapidly iterate on research ideas. Any time we wanted to experiment with a new Chemprop feature, we would have to add a parameter to several functions and classes and trace its path through all of the Chemprop code which could take some time to get right. So for the purpose of rapidly iterating on research ideas, we switched to the design of using a single args object and passing it around. That way, adding a new feature was as simple as adding a new argument and then it would immediately be accessible everywhere throughout Chemprop.

Since Chemprop is still being used actively in research projects, we've continued to opt for the args design for maximum flexibility. However, I totally agree that the explicit design you've presented would be superior for using Chemprop as is since it would make it more modular and easier to fit into other workflows.

I'm curious, do you have thoughts about the right tradeoff between the ease of development using the args approach versus the modularity of the explicit parameter approach? Is there some way to enable both at the same time? I would love to hear what you think about the best way to structure Chemprop to make it both modular and easy to experiment with.

**Comment by davidegraff** [comment]
Hi @swansonk14, thanks for the response!

I'll start off by retracting some of my earlier points about creating wrapped train() and predict() methods. While those are convenient elements of sklearn and tensorflow model objects, pytorch models are fundamentally different in this regard. I don't think this is something you should spend your time on as it's not your job to do so when creating a pytorch-based ML model.

In regards to modularizing the model itself, could you provide a little more detail (to the extent possible, given that you are actively working on it) on what you mean by adding features to the model that made it significantly easier to operate on just one Namespace object? I've spent a bit of time poking through your codebase, and my understanding is that you mostly operate by passing a massive Namespace object around which holds _all_ of the arguments to several, unrelated functions at any one time. 

Absent a clearer idea of your difficulties in using proper function arguments, I might recommend using the `**kwargs` construct in your code. You can instantiate your Namespace object as normal, but then convert this to a dictionary using the `vars` builtin. Instead of calling all of your functions with a monolithic Namespace argument, you would then convert all of your functions to using a monolithic dictionary, `**kwargs`. At least in this construct, it is significantly more straightforward to refactor your code to use proper arguments when testing is completed, because every entry accessed in the kwargs dictionary inside a given function should reasonably be an argument to that function. It also maintains the flexibility of the args approach because the newly added parameter (which may later be removed) resides inside the kwargs dictionary that is passed from function to function.

I'm not sure if this quite addresses your point, but I'd be happy to talk more about this to get a clearer picture of what you mean! Thanks again and great work!

**Comment by swansonk14** [comment]
Hi @davidegraff, thank you for the quick response and for the great discussion!

I think a good example of the benefit of the (admittedly ugly) monolithic namespace approach is when we attempted to test an idea based on label propagation back in March, which involved essentially creating two Chemprop networks that worked simultaneously. You can see the diff between the branch containing that code and the master branch here: https://github.com/chemprop/chemprop/compare/master...chemprop:label_propagation. The label propagation approach required adding a few new arguments to `parsing.py` (which is now `args.py` in the most recent version of Chemprop), and some of these arguments needed to be used at several points throughout Chemprop. For instance, `args.similarity_network` is used in `run_training.py`, `evaluate.py`, and `utils.py`. Since all three of these locations had functions that took an args object as input, we then had instantaneous access to `args.similarity_network` in all of these locations without the need to add `similarity_network` as a parameter in each of these function signatures. This made it possible to build (and subsequently reject) the label propagation idea more quickly than if we had had to change a bunch of function signatures for each of the new label propagation arguments.

I think the `**kwargs` approach is a good way to transition towards fully explicit parameters, but in the intermediate it might lose some of the interpretability of the current approach. If we were still using argparse's `Namespace`, which is essentially a blob of args with no defined structure, then I would agree that `**kwargs` and `Namespace` are essentially equivalent since in both cases you don't know what's inside. However, partly to address this issue, I refactored Chemprop to instead use the [Typed Argument Parser](https://github.com/swansonk14/typed-argument-parser) (Tap) package. Since Tap objects are classes with explicit class variables, unlike `Namespace` objects, an IDE like PyCharm can statically check which arguments are in the Tap object and can do things like static type checking and code completion. Although it won't tell you which arguments are used in any particular function, it at least provides some measure of certainty about what the Tap object contains, whereas this would not be the case with `**kwargs`.

The closest I've come to some sort of compromise between monolithic args objects and explicit parameters is in functions like `get_data` (https://github.com/chemprop/chemprop/blob/master/chemprop/data/utils.py#L98), where the function takes as input both explicit parameters and an args object, and all are optional. That way, a user who wants to use import the function and use it without an args object can use the explicit parameters, while a researcher can instead use the args object without passing any explicit parameters. As the researcher develops a new method, they initially only rely on the args object and add whatever new arguments then want to use in `get_data`, and then once they feel confident that a new argument is work keeping, they then make it explicit in the function signature. Internally, this makes the `get_data` function pretty messy since it needs to deal with the possibility of getting either a single args object or explicit parameters (or possibly both), but it provides maximal flexibility to the user who can decide how to use the function.

Do you think that sort of construction is reasonable, at least from the user perspective? And if so, is there any way to simplify the internals of `get_data` to more easily accommodate either an args object or explicit parameters? Right now it requires a line of code for each argument in order to select between the args object and the explicit parameter, but maybe there's a simpler way to do this? And if you don't think this is a good solution, are there other possibilities that could simultaneously support args objects and explicit parameters? Or do you think explicit parameters are really the best way to go and we should minimize use of the args objects where possible?

I've been struggling with these tradeoffs for a while so I would really appreciate hearing your thoughts!

**Comment by davidegraff** [comment]
You definitely have a valid point here, and I want to stress that a lot of this comes down to personal opinion so I appreciate you giving me feedback some thought! I definitely see the 'ease' of using a single, monolithic `Namespace` for quickly testing out new variables/parameters across several functions. I know how cumbersome it can be to change function signatures for several functions inside a small, and the inconvenience is only exacerbated in a larger library such as yours. I have a few thoughts on your detailed response, but please disregard them if you disagree. It's your library, and you've done a great job already!

1. We agree that there is functionally no difference between a standard `Namespace` and a `**kwargs` approach. My point is that with the `**kwargs` approach, even if you leave all of your function signatures as `def func(**kwargs)`, a user can inspect your codebase for what keyword arguments a function calls and use them in their own function calls. Here is an example of what I mean:

```python
# Chemprop library code
def kwargs_add(**kwargs):
    a = kwargs['a']
    b = kwargs['b']
    return a + b

# end user's function
def my_add(a, b):
    return chemprop_add(a=a, b=b)
```
In principal, there's no difference between the user just adding an extra `Namespace(...)` wrapper around their function call of your library code (like I do in my example at the top,) but it's just slightly more cumbersome and, in essence, muddies up another user's library because of your decision. I don't like to level judgements of "poor style" at other people's code, but your own stylistic decisions in chemprop now potentially affect other user's libraries by forcing them to write 'un-pythonic' code. I want to stress that this isn't a big deal, but it is something for you to consider in your approach

Using Tap objects to help IDEs with type checking of arguments is certainly a good middle ground with this approach.

2. I like `get_data` as a compromise between explicit parameters and using the `Namespace` approach. It definitely is better than the alternative of having some arguments used in a function obscured inside a `Namespace` object. The drawback, though, of offering this flexibility is that an end-user has no ability to rely on default arguments in this function. Because as you've written it, the defaults reside inside the `args` argument, but I find it reasonable that an end-user might reasonably want to avoid this in their code in favor of explicit arguments and implicit defaults. Using your library to generate a proper `Namespace` object that contains these defaults gets back to my original point: that the `Namespace` generated will contain a bunch of attributes of which only a select few are relevant in any given function.

An aside, this is the only way I can think of refactor the code block at the top of `get_data`, and I don't like it very much.
```python
def get_data(...):
    # some code here
    if args is not None:
        # Prefer explicit function arguments but default to args if not provided
        smiles_column = check_explicit(smiles_column, args, 'smiles_column')
        target_columns = check_explicit(target_columns, args, 'target_columns')
        # more calls to check_explicit()
    # rest of the function

def check_explicit(explicit, args, keyword):
    return getattr(args, keyword) if explicit is None else explicit
```

3. I think explicit parameters are firmly the way to go when you release Chemprop 1.0. As it stands right now, it's almost impossible to fully document your API because of the use of `Namespace` objects. You could circumvent this by documenting what attributes a given `Namespace` might need when used as an argument for a particular function, but at that point, you're working so hard to force an unnatural function signature it would honestly be easier to convert everything to an explicit function signature. I can't think of any libraries that use `Namespace`s in place of named arguments.

Happy to continue this conversation and hear your perspective on these things!

**Comment by swansonk14** [comment]
Hi @davidegraff, thank you for the great discussion and sorry for the long delay in replying! I had to prioritize some other work for a while unfortunately. I agree with a lot of what you're saying but I guess ultimately I'm not super happy with either the `**kwargs` solution or the `Tap` solution because it really feels like explicit parameters are the best.

Maybe an alternate solution is to make all the parameters fully explicit in the master branch, and then researchers can simply create their own branches and add in the `Tap` object as a parameter wherever they feel like they need greater flexibility while doing experimentation. Then once the new feature is ready to be merged into master, the researcher can turn the `Tap` arguments that they want to keep into explicit parameters and all the other arguments can be removed. That might be a reasonable middle ground so that the public facing master branch of Chemprop is the most stable and well documented while research branches can be flexible as needed.

If you would be willing to help make some of the functions explicit that would be amazing, but no pressure! I'll try to work on this myself over the coming weeks, but it may take some time as I do have some other projects that I need to focus on as well. However, I really do appreciate all the feedback you've given so I would love to incorporate some of it to improve Chemprop as a package!

**Comment by davidegraff** [comment]
this is being handled in the refactor
